src/cls/rbd/cls_rbd.cc

```diff
@@ -2109,6 +2115,140 @@ int object_map_update(cls_method_context_t hctx, bufferlist *in, bufferlist *out
   return r;
 }
 
+
+static const string metadata_key_for_name(const string &name)
+{
+  return RBD_METADATA_KEY_PREFIX + name;
+}
+
+static const string metadata_name_from_key(const string &key)
+{
+  return key.substr(strlen(RBD_METADATA_KEY_PREFIX));
+}
+
+/**
+ * Input:
+ * @param in ignored
+ *
+ * Output:
+ * @param value
+ * @returns 0 on success, negative error code on failure
+ */
+int metadata_list(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
```

Comment:
since we're only listing up to 1024 keys, metadata_set should probably return an error if there are already 1024 keys (ERANGE to not conflict with other uses of ENOSPC?)